Add maintenance mode to run controller (#204)
SAP/stewardci-core: docs/upgrade/README.md

# Upgrade Guide

To upgrade steward you can set the system in the 'upgrade mode'.
In this mode running pipeline runs will be processed to the end.
Newly created pipeline runs will stay untouched.

## Steps

> - Set the system in the 'upgrade mode'

Reviewer: ```suggestion
- Put Steward in 'upgrade mode'
```

> - Wait until the already started pipeline runs are finished.

Reviewer: ```suggestion
- Wait until already started pipeline runs are finished.
```

- Install the new version of steward

> - Switch off 'upgrade mode'

Author: Done

### Set the system in the 'upgrade mode'

To set the system in the 'upgrade mode' you need to create a config map in the system namespace with specific content.
See the file 'upgrade_mode_on.yaml' in this directory as an example.

```bash
kubectl apply -n steward-system -f upgrade_mode_on.yaml
```

## Wait until the already started pipeline runs are finished.

> You can list all non-finished pipelines with the following command

Reviewer: ```suggestion
You can list all non-finished pipelines with the following command:

```

```bash
 kubectl get spr --all-namespaces | grep -v finished
 ```

> Installation can start when there are none or only pipeline runs with state unknown (empty string).

Reviewer: ```suggestion
The installation can start when there are no pipeline runs at all or _all_ pipeline runs are in one if these states:

- <blank>
- new
- finished
```

Author: I think pipeline runs in state new will be processed...

Reviewer: Proposed again with change.

> For each pipeline run with unknown status you can also see an event with Reason 'SkipOnMaintenanceMode'

Reviewer: ```suggestion
For each pipeline run without state you can also see an event with reason 'SkipOnMaintenanceMode':
```

> For each pipeline run with unknown status you can also see an event with Reason 'SkipOnMaintenanceMode'

Reviewer: How do we call it, "maintenance mode" or "update mode"? Decide for one and use it consistently.

Author: maintenance mode

```bash

> TENANT_

Reviewer: There's something missing.

Reviewer: Proposed again on latest revision.

> kubectl get event -n  $TENANT_NAMESPACE

Reviewer: Better quote expansions:
```suggestion
kubectl get event -n  "$TENANT_NAMESPACE"
```

LAST SEEN   TYPE     REASON                  OBJECT                 MESSAGE
12s         Normal   SkipOnMaintenanceMode   pipelinerun/ok-n9lcl   Maintenance mode skip
```

> ### Install the new version of steward

Reviewer: ```suggestion
### Install the new version of Steward

```

See [Installing Steward](../install/README.md)

### Switch off 'upgrade mode'

> To switch off the upgrade mode you can either delete the created config map or set `upgradeMode: "false"`
> 
> ```bash
> kubectl apply -n steward-system -f upgrade_mode_off.yaml
> ```
> 
> ```bash
> kubectl delete -n steward-system -f upgrade_mode_on.yaml
> ```

Reviewer: ````suggestion
To switch off the upgrade mode you can either delete the created config map:

```bash
kubectl delete -n steward-system -f upgrade_mode_on.yaml
```

or set `upgradeMode: "false"`:

```bash
kubectl apply -n steward-system -f upgrade_mode_off.yaml
```
````
